File: api/app/schemas/shopping.py
import uuid

from pydantic import BaseModel, Field


class ShoppingAdd(BaseModel):
    slug: str
    #: Servings to buy for. Defaults to the recipe's base yield.
    target_yield: int | None = Field(default=None, ge=1)


class ShoppingDelete(BaseModel):
    """Ids to remove in one go, so clearing a selection is a single request."""

    ids: list[uuid.UUID] = Field(min_length=1, max_length=400)


class ShoppingToggle(BaseModel):
    checked: bool


class ShoppingItemOut(BaseModel):
    id: uuid.UUID
    name: str
    amount: float
    unit: str
    #: Rendered by services/scaling.format_amount, so quantities read the same here
    #: as everywhere else — kitchen fractions, em dash for to-taste.
    display: str
    to_taste: bool
    checked: bool
    #: Slugs that contributed to this line.
    recipes: list[str] = []
    #: True for ingredients that routinely hide gluten — worth verifying in the shop.
    check_gluten: bool = False
    #: Which part of the shop this comes from, so the list can be walked once.
    aisle: str = "Other"


class ShoppingListOut(BaseModel):
    items: list[ShoppingItemOut] = []

File: ios/TheSharpEdge/Models/Shopping.swift
import Foundation

/// Mirrors api/app/schemas/shopping.py.
///
/// Quantities are merged server-side: adding a second recipe adds to a line rather
/// than replacing it, and the arithmetic (unit families, kitchen fractions) is the
/// canonical implementation in services/shopping.py. The client only displays.

struct ShoppingItem: Codable, Identifiable, Hashable {
    var id: UUID
    var name: String
    var amount: Double
    var unit: String
    /// Already rendered by the server — kitchen fractions, "to taste" for amount 0.
    var display: String
    var toTaste: Bool
    var checked: Bool
    var recipes: [String]
    /// Routinely hides gluten; worth verifying the label in the shop.
    var checkGluten: Bool

    enum CodingKeys: String, CodingKey {
        case id, name, amount, unit, display, checked, recipes
        case toTaste = "to_taste"
        case checkGluten = "check_gluten"
    }
}

struct ShoppingList: Codable {
    var items: [ShoppingItem] = []
}

struct ShoppingAdd: Encodable {
    var slug: String
    var targetYield: Int?

    enum CodingKeys: String, CodingKey {
        case slug
        case targetYield = "target_yield"
    }
}

struct ShoppingToggle: Encodable {
    var checked: Bool
}
